content/post/2021-07-29-caching/index.Rmd

```diff
@@ -0,0 +1,223 @@
+---
+slug: cache
+title: "How to cache results in R code" 
+authors: 
+- Maëlle Salmon 
+- Christophe Dervieux
+date: "2021-07-29" 
+tags: 
+- package development 
+output: hugodown::hugo_document
+---
+
+```{r setup, include=FALSE}
+
+knitr::opts_chunk$set(fig.path = "", comment = "")
+# knitr hook to make images output use Hugo options
+knitr::knit_hooks$set(
+  plot = function(x, options) {
+    hugoopts <- options$hugoopts
+    paste0(
+      "{{<figure src=",
+      '"', x, '" ',
+      if (!is.null(hugoopts)) {
+        glue::glue_collapse(
+          glue::glue('{names(hugoopts)}="{hugoopts}"'),
+          sep = " "
+        )
+      },
+      ">}}\n"
+    )
+  }
+)
+
+```
+
+One principle of programming that's often encountered is "DRY", "Don't Repeat Yourself", that encourages e.g. the use of functions other duplicated (read: copy-pasted and slightly amended) code.
+You could also interpret it as don't let the machine repeat its calculations if useless.
+How about for a piece of code (function) with the same inputs, we only run it once per R session, and save the results for later?
+In this post, we shall go over ways to cache results in an R session, so that you don't need to burden machines.
+We will not cover [caching for R Markdown](https://bookdown.org/yihui/rmarkdown-cookbook/cache.html).
+
+## Caching: what is it and why use it?
+
+Caching means that if you call a function several times with the exact same input, the function is only actually run the first time.
+The result is stored in a cache of some sort (more practical details later!).
+Every other time the function is called with the same input, the result is retrieved from the cache.
+You will often think of caching as something valid in only one R session, but we'll see it can last longer via storage on disk.
```

Comment:
>  we'll see it can last longer via storage on disk.

For now, there is not really mention later of storage on disk ? 

Should we talk more about it ? 

There is only mention in memoise package part about the other post on persistent storage.
https://github.com/r-hub/blog/pull/146/files#diff-478079e75108d46528efbfb77adfe1490a09e967ffef3a06d2e6404a1f1df107R87

Maybe should highlight that more. 

Key point I see: 

* If you want to persist some information / result between session, you need to write somewhere (usually on disk). The next time the package runs, it will try to read the stored value. 
* If you do that, you can't write locally - best practice is user data dir (see rappdirs package or `tools::R_user_dir()` in R 4.0+
* About formats, text file is great for short string value. Writing compressed RDS file is also an option. In any case, cache storage should be small for internal use in the package. (I want to differentiate from user's computation caching that a package like targets could offer.)